### ISSUE TITLE
Raise error while trying to concat two arrays that are unequal

### ISSUE THREAD

**Issue by Sentient07** [Lasagne/Lasagne#851]
When I try to merge two arrays that have different dimension along the axis that is not in consideration for merging, I don't get an error. However the error is raised at the time when that layer is used as `incoming` or when I try to print it's output shape. Shouldn't the `match` function in `get_output_shape_for` be in `__init__` ?

**Comment by f0k** [comment]
> Shouldn't the `match` function in `get_output_shape_for` be in `__init__ `?

The idea is that `get_output_shape_for` can also process other shapes than `self.input_shape`, e.g., if you have a fully-convolutional network with some shapes specified as `None`, you can call `lasagne.layers.get_output_shape` to obtain the output shape for a particular fully-specified input shape. So we'd like to have the check in `get_output_shape_for`, not only in `__init__`.

It's still unfortunate that the error is not raised early, but only when the next layer is added or the output shape is requested. That's a more general problem, not just for the concatenation layer. We could call `self.output_shape` at the end of every constructor to ensure the shapes are sane. We cannot place this call in the super constructor, though, because at that time not all attributes may be set that are required by `get_output_shape_for`. We'd need something else in the base class that's called after instantiating the layer. I think we had a long discussion about this, but I cannot find it quickly (it may have been via email before the github repository existed). Any ideas that only require changes in the base class are welcome!

But yes, we could also have a shape check in the constructor for this specific layer.

**Comment by Sentient07** [comment]
> We could call self.output_shape at the end of every constructor to ensure the shapes are sane

This seems to be an excellent idea to me. 

> because at that time not all attributes may be set that are required by get_output_shape_for.

I can think of only one way for this, by keeping a method that calls `get_output_shape_for` to ensure that the shape of a Layer passed to this particular Layer is sane. But the doubt is where to make that call from. We can add a call at the end of the constructor, but it would become a mundane task to add it to all the Layers that exists

**Comment by f0k** [comment]
> We can add a call at the end of the constructor, but it would become a mundane task to add it to all the Layers that exists

Exactly. It would also require to change all layers of all users, which is something we cannot do. So this is not an option. The other way would be to move the initialization to an `init()` method called from the base layer's `__init__()` method, so the base layer's `__init__()` can call `get_output_shape_for()` after `init()`, but again, this would require changing all layers, and it would be bewildering.

But as I said in the end, we could also add a check only for layers that would clearly benefit from this.

We could also change the base layer constructor to catch exceptions in computing the previous layer's output shape and then give a more meaningful error message, like "We're sorry we didn't notice earlier, but the FooLayer you're trying to add a BarLayer on top of cannot compute its output shape, indicating a problem with FooLayer's configuration or input. The exception was: ...".
